integration/tests/overlay-focus-trap.test.js

```diff
@@ -0,0 +1,140 @@
+      overlay = fixtureSync('<vaadin-overlay focus-trap></vaadin-overlay>');
+      await nextRender();
+      overlay.renderer = (root) => {
+        if (!root.firstChild) {
+          root.innerHTML = `
+            <vaadin-text-field></vaadin-text-field>
+            <vaadin-radio-group>
+              <vaadin-radio-button id="radioButton1" label="Button 1"></vaadin-radio-button>
+              <vaadin-radio-button id="radioButton2" label="Button 2"></vaadin-radio-button>
+              <vaadin-radio-button id="radioButton3" label="Button 3"></vaadin-radio-button>
+            </vaadin-radio-group>
+            <vaadin-button>tabindex 0</vaadin-button>
+          `;
+        }
+      };
+      overlay.requestContentUpdate();
+      focusableElements = getFocusableElements(overlay.$.overlay);
+    });
+
+    afterEach(() => {
+      overlay.opened = false;
+    });
+
+    it('should properly detect focusable elements inside the content', () => {
+      expect(focusableElements.length).to.equal(6);
+      expect(focusableElements[0]).to.eql(overlay.$.overlay);
+      expect(focusableElements[1]).to.eql(overlay.querySelector('vaadin-text-field').focusElement);
+      expect(focusableElements[2]).to.eql(overlay.querySelector('#radioButton1').focusElement);
+      expect(focusableElements[3]).to.eql(overlay.querySelector('#radioButton2').focusElement);
+      expect(focusableElements[4]).to.eql(overlay.querySelector('#radioButton3').focusElement);
+      expect(focusableElements[5]).to.eql(overlay.querySelector('vaadin-button'));
+    });
+
+    it('should focus focusable elements inside the content when focusTrap = true', async () => {
+      overlay.opened = true;
+      await oneEvent(overlay, 'vaadin-overlay-open');
+
+      // Tab
+      for (let i = 0; i < focusableElements.length; i++) {
+        const focusedIndex = getFocusedElementIndex();
+        expect(focusedIndex).to.equal(i);
+        tabKeyDown(focusableElements[focusedIndex]);
+      }
+      expect(getFocusedElementIndex()).to.eql(0);
+
+      // Shift + Tab
+      tabKeyDown(focusableElements[getFocusedElementIndex()], ['shift']);
+      for (let i = focusableElements.length - 1; i >= 0; i--) {
+        const focusedIndex = getFocusedElementIndex();
+        expect(focusedIndex).to.eql(i);
+        tabKeyDown(focusableElements[focusedIndex], ['shift']);
+      }
+      expect(getFocusedElementIndex()).to.eql(focusableElements.length - 1);
+    });
+  });
+
+  describe('elements in shadow DOM', () => {
+    beforeEach(async () => {
+      overlay = fixtureSync('<vaadin-overlay focus-trap></vaadin-overlay>');
+      await nextRender();
+      overlay.renderer = (root) => {
+        if (!root.firstChild) {
+          const div = document.createElement('div');
+          div.attachShadow({ mode: 'open' });
+          div.shadowRoot.appendChild(document.createElement('vaadin-text-field'));
+          div.shadowRoot.appendChild(document.createElement('button'));
```

Comment:
Changed the `input` to `button` here just to simplify the usage of `querySelector` in the test below.
Otherwise there would be two `input` elements including the one created by `vaadin-text-field`.